util/system/path.go

```diff
@@ -5,10 +5,10 @@ package system
 // ':' character .
 const DefaultPathEnvUnix = "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
 
-// DefaultPathEnvWindows is windows style list of directories to search for
-// executables. Each directory is separated from the next by a colon
-// ';' character .
-const DefaultPathEnvWindows = "c:\\Windows\\System32;c:\\Windows"
+// DefaultPathEnvWindows is deliberately empty on Windows and the default path
+// must be set by the image. BuildKit has no context of what the default
```

Comment:
I don't understand where that "must be set by the image" is coming from. Yes, in wcow environment variables can be part of the rootfs tars because of the registry files, but they can also be in image configuration.

I don't see any reason to add these exceptions to the codebase where they need to be understood by the devs and also by people writing the Dockerfiles. To have a reasonable behavior that works with `ENV` in Dockerfiles the images should define `PATH` in the config and users who wish to change it should do it with `ENV` command, the same way as the Linux users do. Yes, there is another method to do it but it doesn't mean it needs to be used.

Reply:
Windows is... different.

First of all, Windows doesn't support `FROM scratch`, so _every Windows image_ will have a "Windows" flavour as base image. It's up to that base image to define the `PATH` (in whatever form it does).

So unless your creating a base image (which means you're either Working for Microsoft, or violating your Windows License), the `PATH` should be defined by the base image, so the "default" is determined by the base image.

Of  course, as an image author you may override that default, but BuildKit itself should not try to come up with a default, because there _isn't_ any.

To be fair, I don't think we should've set a default for non-Windows images either (and also have let it up to the base image (if any)), but that ship sailed. Changing that would be a breaking change, but so is https://github.com/moby/buildkit/pull/1747, which changed the default to be different from _any_ container system created in the past 7 Years).

Reply:
I agree that the base image should set the `PATH`. The default we have should only be a fallback that is almost never used(it is almost never used in Linux as well). People who create Windows images should set the `PATH` in the image config the same way people who create Linux images do, not ask tools to add and maintain a bunch of exceptions for them. This is also confusing for the users when `ENV PATH=$PATH` has unexpected behavior.

Reply:
The issue is that (but don't ask me about details, @TBBle or @kevpar may be able to explain it better), that setting a `PATH` breaks things on Windows.

For example, take the image config for `mcr.microsoft.com/windows/servercore:ltsc2022`;

```bash
docker buildx imagetools inspect --raw mcr.microsoft.com/windows/servercore:ltsc2022@sha256:06bb218927a8a95b424acfad0e0e6ae0fb5694c0b510f9930ed5881075eaec2f | jq .
{
  "architecture": "amd64",
  "config": {
    "Hostname": "",
    "Domainname": "",
    "User": "",
    "AttachStdin": false,
    "AttachStdout": false,
    "AttachStderr": false,
    "Tty": false,
    "OpenStdin": false,
    "StdinOnce": false,
    "Env": null,
    "Cmd": [
      "c:\\windows\\system32\\cmd.exe"
    ],
    "Image": "",
    "Volumes": null,
    "WorkingDir": "",
    "Entrypoint": null,
    "OnBuild": null,
    "Labels": null
  },
  "created": "2022-10-07T22:13:48.3974633Z",
  "history": [
    {
      "created": "2022-04-22T01:12:09.4542389Z",
      "created_by": "Apply image 10.0.20348.643"
    },
    {
      "created": "2022-10-07T22:13:48.3974633Z",
      "created_by": "Install update 10.0.20348.1129"
    }
  ],
  "os": "windows",
  "os.version": "10.0.20348.1129",
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:b9bda40f596a8f8648c218b23e552acf9498c039eb0a48c222cfcd5a9af9e841",
      "sha256:7a72bd51b9c8b0510fea5addf0b46e5de102813083abe316d4603a050e7ad473"
    ]
  }
}
```

When I tried to update the classic builder to use the default `PATH` from this repository, things break. Not _exactly_ sure, but I somewhat expect that in this case either `cmd` or `PowerShell` define the path, but when _setting_ a `PATH` as environment variable, that overrides that default, rendering the image broken;

```bash
Step 1/13 : ARG WINDOWS_BASE_IMAGE=mcr.microsoft.com/windows/servercore
Step 2/13 : ARG WINDOWS_BASE_IMAGE_TAG=ltsc2022
Step 3/13 : ARG BUSYBOX_VERSION=FRP-3329-gcf0fa4d13
Step 4/13 : ARG BUSYBOX_SHA256SUM=bfaeb88638e580fc522a68e69072e305308f9747563e51fa085eec60ca39a5ae
Step 5/13 : FROM ${WINDOWS_BASE_IMAGE}:${WINDOWS_BASE_IMAGE_TAG}
ltsc2022: Pulling from windows/servercore
97f65a0ec59e: Pulling fs layer
4486102fd382: Pulling fs layer
4486102fd382: Verifying Checksum
4486102fd382: Download complete
97f65a0ec59e: Verifying Checksum
97f65a0ec59e: Download complete
97f65a0ec59e: Pull complete
4486102fd382: Pull complete
Digest: sha256:2846453bcfaee661d563e29d6513191dccf7280bb826b2da9d6e7226fba2db94
Status: Downloaded newer image for mcr.microsoft.com/windows/servercore:ltsc2022
 ---> 2caccf6918d4
Step 6/13 : RUN mkdir C:\tmp && mkdir C:\bin
 ---> Running in 320b82116e6d
Removing intermediate container 320b82116e6d
 ---> 5e9277f6f4b4
Step 7/13 : ARG BUSYBOX_VERSION
 ---> Running in a85cdf300ed4
Removing intermediate container a85cdf300ed4
 ---> a0d354994a80
Step 8/13 : ARG BUSYBOX_SHA256SUM
 ---> Running in 99f510599835
Removing intermediate container 99f510599835
 ---> ae8d3ccee635
Step 9/13 : ADD https://frippery.org/files/busybox/busybox-w32-${BUSYBOX_VERSION}.exe /bin/busybox.exe

 ---> 0c2ef53e2a7c
Step 10/13 : RUN powershell     if ((Get-FileHash -Path /bin/busybox.exe -Algorithm SHA256).Hash -ne $Env:BUSYBOX_SHA256SUM) {         Throw \"Checksum validation failed\"     }
 ---> Running in 65d36517920b
'powershell' is not recognized as an internal or external command,
operable program or batch file.
The command 'cmd /S /C powershell     if ((Get-FileHash -Path /bin/busybox.exe -Algorithm SHA256).Hash -ne $Env:BUSYBOX_SHA256SUM) {         Throw \"Checksum validation failed\"     }' returned a non-zero code: 1
```

I guess this is where the (_LONG_ ) discussions we had about https://github.com/moby/moby/pull/29048 come to play; the `PATH` environment variable cannot be read from the environment, but must be read back from the process that's running (e.g. `PowerShell` or `cmd.exe`).

So probably the only options for the builder are;

- Don't set `PATH` and acknowledge that "we don't know"
- Execute the container's process and ask it what the variable is (implement `GETENV`)

Reply:
The breakage above happens because PowerShell lives in `C:\WINDOWS\System32\WindowsPowerShell\v1.0\` which is in-the path on systems with PowerShell present (and hence not on nanoserver). See also https://github.com/microsoft/terminal/issues/7418 for other env-var traps in process launching (again, with PowerShell).

https://github.com/moby/moby/pull/31525#issuecomment-285739019 seems to be the best reference for how we got here. The `ENV` in Dockerfile and the registry-held persistent env-vars (both machine and user) are parallel sources of truth, and the simplest workaround is to just not use `ENV PATH` in Windows Dockerfiles, in favour of `RUN cmd /c setx /m PATH=${PATH};...`. (Or `RUN reg <options that append to a multi-string value>` to save you using `${PATH}`, etc.)

This is most-compatible with installers that add paths to the registry as well. Once you've `ENV PATH`'d once, you then need to `ENV PATH` after each installer executes too as you lose access to all past _and future_ registry-based env-path changes.

That's not a *great* workaround, as it puts extra load on the Dockerfile writer, and requires `RUN`, which blocks non-executable-environment layer-editing tools. That's also true of GETENV as described, although if some brave soul wants to use a hive-loading library, e.g., https://github.com/gabriel-samfira/go-hivex, to implement GETENV (or really env-substitution) without booting a container, that option exists and I think it less-awful than it might seem at-first. (Still not _great_ though).

Along those lines, making Windows-platform builders transform Dockerfile `ENV` into a `RUN setx` instead of writing to the container image config would resolve down to a single source of truth, with the same limitations as above _and_ with extra build-time cost of launching another container for the `ENV` command, which would be surprising to long-time users.

As I understand, there's no good way to tell a Windows process being launched to _append_ an env-var to the registry-provided values, since the Windows process launch takes a single set of env-vars, and it's up to the launching process to assemble them from either copying the existing environment, or [requesting a new environment block from the OS](https://learn.microsoft.com/en-us/windows/win32/api/userenv/nf-userenv-createenvironmentblock) and hence using only the registry (basically only explorer and similar process launchers do the latter, and we can't do it outside a running container anyway). This is where Windows's approach differs from Linux, where if you need to append to the path, you add an `export` to the end of .bashrc, and hence it's trivially additive across container image layers.

A more esoteric solution would be some wild inversion of the above, which I am not currently volunteering to write: At the end of every `RUN`, the Windows registry env-vars are exported into the container config `ENV` block (matching the merge/override behaviours as they are processed in Windows, perhaps with a secret `RUN env` where we don't keep the scratch layer, or using a hive-reading library). That way Dockerfile ENV is the canonical way to make changes, but changes by installers or `setx /m` are correctly captured as well.

[The similar issue around stale env-vars in Windows Terminal tabs](https://github.com/microsoft/terminal/issues/1125) has been around for years without successful resolution. It's slightly more complex because its two sources of truth overlap (one contains an old version of the other, so there's a three-way merge in the mix), where here a fairly simple merge is sufficient.

Reply:
Adding my 2 cents:

There is also the annoying fact that a potential installer may set just the system ```PATH``` or the user ```PATH```...or both. So the ```PATH``` will differ based on what we set in the ```USER``` stanza.

For example:

```Dockerfile
USER ContainerAdministrator
RUN echo %PATH%
USER gabriel
RUN echo %PATH%
```

The two ```RUN``` commands above may yield different results, and that may be something that is expected by the consumers of the image.

Not setting the ```PATH``` as an env var is, I **think**, desirable on Windows.

Reply:
You really don't want to do `setx /m PATH=${PATH};...` either, because that takes expanded value of PATH from the current process environment (where it includes appending per-user values, expanded substitutions of any REG_EXPAND_SZ `%OTHERENVVAR%` references, etc) and puts that whole string back into `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\Environment`. So your per-user stuff has leaked over to other users, any references are now copied-by-value, your own user values will be appended a second time (and thus become duplicated if `setx` is used like this again.